### PR TITLE
dev: update clippy CI

### PR DESCRIPTION
**Pull request by bhavyagosai**
Time spent on this PR:

Related: #584

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

- Updated CI to deny some useful lints
- have also added `clippy::nursery` here but its linting something that i think is a bug in clippy, i am not able to understand it. added it here so someone can take a look, we can remove `clippy::nursery` if its causing those false positives.

# Does this introduce a breaking change?

- [x] Maybe

### CODE REVIEW

**Comment by bhavyagosai** [comment]
@Eikix can you rerun clippy workflow, failure seems unreleated. it should error with `Future`'s warning.

File: crates/test-utils/src/fixtures.rs [hunk 1]
```diff
@@ -5,16 +5,16 @@ use crate::execution::contract::KakarotEvmContract;
...
+    let katana = katana.await;
     let eoa = katana.eoa();
     let contract = eoa.deploy_evm_contract("Counter", ()).await.expect("Failed to deploy Counter contract");
     (katana, contract)
 }
 
 #[fixture]
-#[awt]
 pub async fn erc20(#[future] katana: Katana) -> (Katana, KakarotEvmContract) {
+    let katana = katana.await;
```
Comment (Eikix) [inline crates/test-utils/src/fixtures.rs:17]: Why do we need this? (curious to learn)

File: crates/test-utils/src/fixtures.rs [hunk 2]
```diff
@@ -32,6 +32,7 @@ pub async fn erc20(#[future] katana: Katana) -> (Katana, KakarotEvmContract) {
 
 /// This fixture creates a new test environment on Katana.
 #[fixture]
+#[allow(clippy::future_not_send)]
```
Comment (Eikix) [inline crates/test-utils/src/fixtures.rs:35]: Why do we need this

File: crates/test-utils/src/fixtures.rs (hunk 1 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:17]: not able to see the error on CI, but if you run `./scripts/clippy.sh` you will see some errors:

>  `impl Future<Output = Katana>` doesn't implement `Send`

But `Katana` itself implements `Send` so its `Future` also should. So afaik that is a false positive and I was trying to fix that lint. but even after specifically adding `#[allow]` for that lint it was giving same warning.

File: crates/test-utils/src/fixtures.rs (hunk 2 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:35]: see message is previous question

File: crates/test-utils/src/fixtures.rs (hunk 1 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:17]: there was one more lint that didn't make any sense, i think due to this kind of false positive people in general don't like to enable them in CI

i added `clippy::nursery` just to show you guys, will remove it once you confirm its false positive.

Comment (greged93) [reply crates/test-utils/src/fixtures.rs:17]: This is not needed, you can add the `#[awt]` macro to wait for the katana fixture to resolve.

File: crates/test-utils/src/fixtures.rs (hunk 2 above)
Comment (greged93) [reply crates/test-utils/src/fixtures.rs:35]: please update this based on above comment

File: scripts/clippy.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+cargo clippy "$@" --workspace --all-features --all-targets \
+    -- -D warnings -D future-incompatible -D nonstandard-style \
+    -D rust-2018-idioms -D rust-2021-compatibility -D unused -W clippy::nursery
```
Comment (greged93) [inline scripts/clippy.sh:5]: This is not needed and can be moved in linter.yml

File: crates/test-utils/src/fixtures.rs [hunk 4]
```diff
@@ -5,16 +5,16 @@ use crate::execution::contract::KakarotEvmContract;
 use crate::sequencer::Katana;
 
 #[fixture]
-#[awt]
```
Comment (greged93) [inline crates/test-utils/src/fixtures.rs:8]: why did you remove this?

File: crates/test-utils/src/fixtures.rs [hunk 5]
```diff
@@ -5,16 +5,16 @@ use crate::execution::contract::KakarotEvmContract;
...
 pub async fn counter(#[future] katana: Katana) -> (Katana, KakarotEvmContract) {
+    let katana = katana.await;
     let eoa = katana.eoa();
     let contract = eoa.deploy_evm_contract("Counter", ()).await.expect("Failed to deploy Counter contract");
     (katana, contract)
 }
 
 #[fixture]
-#[awt]
```
Comment (greged93) [inline crates/test-utils/src/fixtures.rs:16]: why did you remove this?

**Review by greged93 — COMMENTED** [review]
I'm not sure why you removed some of the macros from rstest on the fixtures.

File: .github/workflows/linters.yml [hunk 6]
```diff
@@ -46,6 +46,5 @@ jobs:
           components: clippy
       - uses: Swatinem/rust-cache@v2
       - uses: actions-rs/cargo@v1
-        with:
-          command: clippy
-          args: --workspace --all-features --all-targets -- -D warnings
+      - run: >
+          scripts/clippy.sh
```
Comment (Eikix) [inline .github/workflows/linters.yml:50]: why not add the args present in the script in the args key?

**Comment by Eikix** [comment]
I think Nursery is a tricky one:)! Let's try and see if we can make a PR for adding this stricter Clippy ourselves, to limit back and forths, 

What changes we will apply in an in-house Pr in the coming days:
- Move away from clippy script and rather add more args in the CI action
- Add ignored lint rules in clippy.toml
- Add strict rules in clippy.toml 

i think from there it'll be easier for you to contribute in a way that suits us:)

File: crates/test-utils/src/fixtures.rs (hunk 4 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:8]: i thought compiler is getting confused in some macro, so i was trying to remove `#[awt]` which manually awaits the future and did it manually, but that still didnt fix it

tldr; the issue persists with and without `#[awt]` macro

File: crates/test-utils/src/fixtures.rs (hunk 1 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:17]: https://github.com/kkrt-labs/kakarot-rpc/pull/656#discussion_r1400736288

Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:17]: i thought compiler is getting confused in some macro, so i was trying to remove #[awt] which manually awaits the future and did it manually, but that still didnt fix it

tldr; the issue persists with and without #[awt] macro

File: crates/test-utils/src/fixtures.rs (hunk 4 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:8]: https://github.com/kkrt-labs/kakarot-rpc/pull/656#discussion_r1400738718

File: crates/test-utils/src/fixtures.rs (hunk 5 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:16]: https://github.com/kkrt-labs/kakarot-rpc/pull/656#discussion_r1400738718

File: crates/test-utils/src/fixtures.rs (hunk 1 above)
Comment (bhavyagosai) [reply crates/test-utils/src/fixtures.rs:17]: but i will readd to make it what it was previously

**Comment by bhavyagosai** [comment]
okay so for now:
- i will remove remove the script and add directly to `linter.yml` and in githooks 
- remove `clippy::nursery` lint

**Comment by bhavyagosai** [comment]
> I think Nursery is a tricky one:)! Let's try and see if we can make a PR for adding this stricter Clippy ourselves, to limit back and forths,
> 
> What changes we will apply in an in-house Pr in the coming days:
> 
> * Move away from clippy script and rather add more args in the CI action
> * Add ignored lint rules in clippy.toml
> * Add strict rules in clippy.toml
> 
> i think from there it'll be easier for you to contribute in a way that suits us:)

okay so for now:

- i will remove remove the script and add directly to `linter.yml` and in githooks
- remove `clippy::nursery` lint

File: .github/workflows/linters.yml (hunk 6 above)
Comment (bhavyagosai) [reply .github/workflows/linters.yml:50]: since we need the exact same command in githooks as well i was trying to reduce duplication, but will revert back to using what you suggested

**Comment by Eikix** [comment]
> > I think Nursery is a tricky one:)! Let's try and see if we can make a PR for adding this stricter Clippy ourselves, to limit back and forths,
> > What changes we will apply in an in-house Pr in the coming days:
> > 
> > * Move away from clippy script and rather add more args in the CI action
> > * Add ignored lint rules in clippy.toml
> > * Add strict rules in clippy.toml
> > 
> > i think from there it'll be easier for you to contribute in a way that suits us:)
> 
> okay so for now:
> 
> * i will remove remove the script and add directly to `linter.yml` and in githooks
> * remove `clippy::nursery` lint
> 
> so these changes can be merged

Sure, let's see if that works

**Comment by bhavyagosai** [comment]
@Eikix CI failure don't look related to this PR as i can see it failing on other PR too

**Comment by Eikix** [comment]
> @Eikix CI failure don't look related to this PR as i can see it failing on other PR too

Investigating

**Comment by bhavyagosai** [comment]
looks like linter.yml file is removed, will remove it form this PR as well to resolve conflict.

**Comment by bhavyagosai** [comment]
again failure unrelated to change

**Comment by bhavyagosai** [comment]
hey so should i merge this with main and fix conflicts or is this no longer needed? @Eikix

**Comment by Eikix** [comment]
Rebase into merge looks fine to me:)

**Comment by bhavyagosai** [comment]
@Eikix done!

**Review by Eikix — APPROVED** [review]
lgtm
